ledger-api/grpc-definitions/com/daml/ledger/api/v1/events_service.proto

```diff
@@ -0,0 +1,364 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+
+/*
+Satisfied Design goals:
+- can co-exist with existing API services for single-domain participants
+- can be served with the 1.17 IndexDB
+- supports serving a snapshot and delta on the same stream
+- serves transaction nodes in individual messages
+- supports retrieving the transaction of a transaction node
+- supports consumers in building the transaction tree
+
+WIP:
+- supports TransactionFilter
+- provides clear docs for how the events are filtered
+- minimize differences to existing transaction_service and active_contracts_service
+- ensure that there is an efficient strategy for serving the stream
+
+Next design goals:
+- support evolving the interface to v2 multi-domain transactions
+- include completions
+- communicate pruning events
+- include all data required for Ledger API completeness
+
+
+*/
+
+
+
+syntax = "proto3";
+
+package com.daml.ledger.api.v1;
+
+import "google/protobuf/wrappers.proto";
+import "com/daml/ledger/api/v1/transaction_filter.proto";
+
+
+option java_outer_classname = "EventsServiceServiceOuterClass";
+option java_package = "com.daml.ledger.api.v1";
+option csharp_namespace = "Com.Daml.Ledger.Api.V1";
+
+
+service EventsService {
+  // Get the address of a recent transaction as of which an event source can be provided.
+  rpc GetRecentSnapshotOffset(GetRecentSnapshotOffsetRequest) returns (GetRecentSnapshotOffsetResponse);
+
+  // Stream events while first serving all events required to reconstruct a state snapshot as
+  // of a given transaction, and afterwards streaming changes to that state.
+  rpc StreamEvents (StreamEventsRequest) returns (stream StreamEventsResponse);
+}
+
+// Offsets
+//////////
+
+// A participant-wide address of a transaction and the state it may depend on.
+message TransactionOffset {
+  // The Ledger API v1 offset
+  oneof kind {
+    // Compared in lexicographic order.
+    string transaction_log_offset = 1;
+
+    // TODO: what kind of type do we need to support multi-domain transactions?
+    // Could the below work?
+    TransactionGraphOffset transaction_graph_offset = 2;
+  }
+}
+
+message TransactionGraphOffset {
+  // Party-id to record_time vector clock.
+  map<string, uint64> vector_clock = 1;
+}
+
+// A participant-wide address of an event and the state it may depend on.
+// Comparisons are done in lexicographic order of the fields.
+message EventOffset {
+  TransactionOffset transaction_offset = 1;
+  LocalEventOffset local_event_offset = 2;
+}
+
+// A transaction-local address of an event.
+message LocalEventOffset {
+  // Equivalent to v1 node-id within a transaction.
+  // TODO: do we need something more powerful?
```

Comment:
How stable do you intend these local offsets to be across parties? Suppose there's a transaction `tx` with nodes 1, 2, 3 where nodes 1 and 3 are visible to `Alice` and `1, 2` are visible to Bob. Ideally, Alice shouldn't learn that there is exactly one node `2` that is not visible to her. So on Alice's event stream, we should have the local offsets 1 and 2, and on the stream for Alice and Bob together we have three local offsets 1, 2, 3.